### ISSUE TITLE
Q: What is the 'best' approach for conditional Task loop in a WorkflowDef? 

### ISSUE THREAD

**Issue by mashurex** [Netflix/conductor#35]
Scenario:
If task Y matches a certain criteria I want to rerun a series of Workflow Tasks from previous step X.
`task_1`: Some type of task that doesn't need to be re-run
`task_2`:Input that influences the proceeding decision
`task_3`: Run processes on` task_2` output that will calculate success/failure status.
`decion_task`: `task_3.output.success == true`, move to `task_4` else go back to `task_2` for retry
`task_4` to `task_n`: All steps that should only be run after the decision task evaluates to `success == true`

Effectively a `do ... while` loop on`task_2` and `task_3`. All of the examples I see for decisions have very concrete flows that don't involve re-running a previous task and then continuing on a single success path when the criteria is met.

In this case `task_2` is complete and not failed because it's process actually worked, but was provided bad input. `task_3` Evaluates the after-effects of `task_2` for validity and passes it's output to the decision task.

Am I better off with a custom task that will influence the decider or executor, or does the decider have the ability to do a loop back via WorkflowDef already?

**Comment by v1r3n** [comment]
Conductor does not support an explicit ```do ... while``` loop as yet.  However, it is possible to execute a task multiple times.   Upon execution of a task by worker, it can return back with either of the three statuses:

1. COMPLETED
2. FAILED
3. IN_PROGRESS

When the task is marked as IN_PROGRESS it is put back in the queue and delivered to the worker again.  However, the worker can set ```callbackAfterSeconds``` field along with the task status to indicate how many second, the server should wait before delivering the task again.  With this, Conductor allows you to execute the same task multiple times with a delay between subsequent execution.

Back to your use case, if you combine ```task_2``` and ```task_3``` in a single task, where the task is marked as COMPLETED only upon meeting the condition.  ```task_{4...n}``` will proceed as normal then.

We have considered looping semantics - but not prioritized yet.  Let us know if you have more use cases, it will help is putting it in our roadmap.

**Comment by mashurex** [comment]
Thanks for the reply @v1r3n. I assumed this was the case but I just wanted to make sure that I wasn't missing something. I've basically done as you said by making a subWorkflow out of `task_2` and `task_3` so that it's easier to go back and restart. I might even create a `rollback` task or something like that that would delete the tasks between itself and it's target task to effectively reset the state of the workflow, but then you lose historical record unless you keep a shadow data store.

**Comment by v1r3n** [comment]
on the historical records, we have this feature that I plan to handle next.  https://github.com/Netflix/conductor/issues/25
Essentially, the idea is to be able to allow tasks to store extended data including history of each execution attempt in a separate store (separate as to avoid overloading the IO during workflow evaluation and potentially be able to store really large amount of data)

**Comment by akshat-g** [comment]
@v1r3n - I created a simple workflow with 3 sequential simple tasks - task_1, task_2, task_3. I started the workflow, and updated the status of task_1 and task_2 as COMPLETED. Now task_3 is in scheduled state. I then tried to update the status of first task task_1 as IN_PROGRESS by hitting POST /tasks api. However task_1 did not go to IN_PROGRESS state. It remained in COMPLETED state. task_3 remained in SCHEDULED state.

This is the payload I used to update status of task_1 from COMPLETED to back to IN_PROGRESS - 

{
"workflowInstanceId": some-workflowId,
"taskId": "taskId of task_1",
"workerId": "1",
"status": "IN_PROGRESS"
}

I would like to know if I am missing anything. Thanks
